fix python binding

pca006132 opened the pull request:
Fixes https://github.com/elalish/manifold/issues/947#issuecomment-2399029206

pca006132 commented:
Not entirely sure why the test is failing. May need a docker to test this.

pca006132 commented:
I guess `warp_batch` was broken... will see how to fix that.

pca006132 commented:
In fact, I am not sure if we should keep the `warp_batch` API. I think this kind of modifying array by reference is quite brittle with transfer of ownership. Returning a new array will always work, and I don't think that will add too much overhead.

elalish commented on pyproject.toml:
```diff
@@ -26,7 +26,7 @@ dependencies = [
 [build-system]
 requires = [
-    "nanobind>=1.8.0",
+    "nanobind>=1.8.0,<=2.2.0",
```
Do we need this if we upgraded to 2.2 above?

pca006132 replied:
Just to be safe. I don't think they provide a stable API for returning ndarray without copy. Maybe we should just ask the upstream.

elalish replied:
Will it matter anymore if we switch to returning a new array?

pca006132 replied:
Still matters, because we still want to return arrays without copying. Returning a new array just avoids reading back the array after passing it to the user, in case numpy did some weird things to the array.

elalish replied:
Okay, yeah let's ask upstream. I think I'd rather eat the copy cost than rely on a non-stable API.

pca006132 replied:
Looking at the code, it seems that we are doing a copy anyway... Maybe we can find a simpler way.

elalish reviewed (commented):
Returning a new array sounds good to me.

pca006132 commented:
Summary:

1. No longer depends on unstable API. I found the solution to be surprisingly simple: just use the type caster for ndarray, it automatically performs conversion from ndarray to python handle (which was why we were trying to call `ndarray_wrap`).
2. Changed the `warp_batch` API to require returning an array. I feel that we were lucky the previous code works, because it is probably relying on numpy not changing the underlying buffer for some operations.
3. Added docstring override functionality, so we can customize some docstring. I think we may also want to add replacement function later, for example to replace different function names due to differences in naming convention.
4. Download 2.2.0 of nanobind, and set the version string if we download it using `FetchContent` (nanobind is unavailable in python search path, so we cannot get the version string automatically in this case).
5. Limit nanobind dependency to 2.2.0. I don't think we relied on unstable APIs now, but to avoid potential future build failure, it is better to specify the latest version that we support. As long as we have regular releases, we can still update nanobind version.
